src/main/java/com/minecolonies/coremod/colony/buildings/BuildingWareHouse.java

```diff
@@ -0,0 +1,230 @@
+     */
+    private static final String TAG_DELIVERYMAN = "Deliveryman";
+
+    /**
+     * The list of deliverymen registered to this building.
+     */
+    private static final List<Vec3d> registeredDeliverymen = new ArrayList<>();
+
+    /**
+     * The tileEntity of the building.
+     */
+    private TileEntityWareHouse tileEntity;
+
+    /**
+     * Instantiates a new warehouse building.
+     *
+     * @param c the colony.
+     * @param l the location
+     */
+    public BuildingWareHouse(final Colony c, final BlockPos l)
+    {
+        super(c, l);
+    }
+
+    /**
+     * Register deliveryman with the warehouse.
+     * @param buildingWorker the building of the worker.
+     * @return true if able to register or already registered
+     */
+    public boolean registerWithWareHouse(BuildingDeliveryman buildingWorker)
+    {
+        if(registeredDeliverymen.contains(new Vec3d(buildingWorker.getID())))
+        {
+            return true;
+        }
+
+        if(registeredDeliverymen.size() >= getBuildingLevel())
+        {
+            if(!registeredDeliverymen.isEmpty())
+            {
+                Log.getLogger().info(getColony().getName() + " " + Arrays.toString(registeredDeliverymen.toArray()));
+                checkForRegisteredDeliverymen();
+            }
+            return false;
+        }
+
+        registeredDeliverymen.add(new Vec3d(buildingWorker.getID()));
+        return true;
+    }
+
+    /**
+     * Check the registered deliverymen and see if one of their huts got destroyed.
+     */
+    private void checkForRegisteredDeliverymen()
+    {
+        final List<Vec3d> registeredDeliverymenCopy = new ArrayList<>(registeredDeliverymen);
+        for(final Vec3d pos: registeredDeliverymenCopy)
+        {
+            final Colony colony = getColony();
+            if(colony != null && colony.getWorld() != null
+                    && (!(colony.getWorld().getBlockState(new BlockPos(pos)) instanceof BlockHutDeliveryman) || colony.isCoordInColony(colony.getWorld(), new BlockPos(pos))))
+            {
+                registeredDeliverymen.remove(pos);
+            }
+        }
+    }
+
+    /**
+     * Check if deliveryman is allowed to access warehouse.
+     * @param buildingWorker the building of the deliveryman.
+     * @return true if able to.
+     */
+    public boolean canAccessWareHouse(BuildingDeliveryman buildingWorker)
+    {
+        return registeredDeliverymen.contains(new Vec3d(buildingWorker.getID()));
+    }
+
+    /**
+     * Get the deliverymen connected with this building.
+     * @return the unmodifiable List of positions of them.
+     */
+    public List<Vec3d> getRegisteredDeliverymen()
+    {
+        return Collections.unmodifiableList(registeredDeliverymen);
+    }
+
+    @NotNull
+    @Override
+    public String getSchematicName()
+    {
+        return WAREHOUSE;
+    }
+
+    @Override
+    public int getMaxBuildingLevel()
+    {
+        return 5;
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Assign this magic number 5 to a well-named constant, and use the constant instead. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS109)